launch/launch/actions/execute_local.py

```diff
@@ -0,0 +1,662 @@
+import traceback
+from typing import Any  # noqa: F401
+from typing import Callable
+from typing import cast
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Text
+from typing import Tuple  # noqa: F401
+from typing import Union
+
+import launch.logging
+
+from osrf_pycommon.process_utils import async_execute_process
+from osrf_pycommon.process_utils import AsyncSubprocessProtocol
+
+from .emit_event import EmitEvent
+from .opaque_function import OpaqueFunction
+from .timer_action import TimerAction
+
+from ..action import Action
+from ..conditions import evaluate_condition_expression
+from ..descriptions import Executable
+from ..event import Event
+from ..event_handler import EventHandler
+from ..event_handlers import OnProcessExit
+from ..event_handlers import OnProcessIO
+from ..event_handlers import OnProcessStart
+from ..event_handlers import OnShutdown
+from ..events import matches_action
+from ..events import Shutdown
+from ..events.process import ProcessExited
+from ..events.process import ProcessIO
+from ..events.process import ProcessStarted
+from ..events.process import ProcessStderr
+from ..events.process import ProcessStdin
+from ..events.process import ProcessStdout
+from ..events.process import ShutdownProcess
+from ..events.process import SignalProcess
+from ..launch_context import LaunchContext
+from ..launch_description import LaunchDescription
+from ..launch_description_entity import LaunchDescriptionEntity
+from ..some_actions_type import SomeActionsType
+from ..some_substitutions_type import SomeSubstitutionsType
+from ..substitution import Substitution  # noqa: F401
+from ..substitutions import LaunchConfiguration
+from ..substitutions import PythonExpression
+from ..utilities import create_future
+from ..utilities import is_a_subclass
+from ..utilities import normalize_to_list_of_substitutions
+from ..utilities import perform_substitutions
+
+_global_process_counter_lock = threading.Lock()
+_global_process_counter = 0  # in Python3, this number is unbounded (no rollover)
+
+
+class ExecuteLocal(Action):
+    """Action that begins executing a process on the local system and sets up event handlers."""
+
+    def __init__(
+        self,
+        *,
+        process_description: Executable,
+        shell: bool = False,
+        sigterm_timeout: SomeSubstitutionsType = LaunchConfiguration(
+            'sigterm_timeout', default=5),
+        sigkill_timeout: SomeSubstitutionsType = LaunchConfiguration(
+            'sigkill_timeout', default=5),
+        emulate_tty: bool = False,
+        output: Text = 'log',
+        output_format: Text = '[{this.process_description.final_name}] {line}',
+        log_cmd: bool = False,
+        on_exit: Optional[Union[
+            SomeActionsType,
+            Callable[[ProcessExited, LaunchContext], Optional[SomeActionsType]]
+        ]] = None,
+        respawn: bool = False,
+        respawn_delay: Optional[float] = None,
```

Comment:
@roger-strain @mlanting meta: looking at this with fresh eyes, none of these arguments (from line 92 to to line 100) are specific to the local execution method. The line is a bit blurry though. Perhaps, these could be part of the process description or at least be pushed to an abstract base class for all execution actions. 

I'm fine with postponing that decision to a follow-up PR though.

Reply:
I would prefer to postpone this for now. Our plan is to add an ExecuteRemote class to support multi-machine launching once we get these changes through. That process should clarify how a base class should look and we can include it in that PR.